packages/ra-ui-materialui/src/input/DateInput.tsx

```diff
@@ -35,42 +35,120 @@ export const DateInput = ({
     defaultValue,
     format = getStringFromDate,
     label,
-    name,
     source,
     resource,
     helperText,
     margin,
     onBlur,
     onChange,
+    onFocus,
     parse,
     validate,
     variant,
     ...rest
 }: DateInputProps) => {
-    const { field, fieldState, formState, id, isRequired } = useInput({
+    const {
+        field,
+        fieldState: { error, invalid, isTouched },
+        formState: { isSubmitted },
+        id,
+        isRequired,
+    } = useInput({
         defaultValue,
-        name,
-        format,
-        parse,
         onBlur,
-        onChange,
         resource,
         source,
         validate,
         ...rest,
     });
+    const [renderCount, setRenderCount] = React.useState(1);
+
+    const initialDefaultValueRef = React.useRef(field.value);
+
+    React.useEffect(() => {
+        const initialDateValue =
+            new Date(initialDefaultValueRef.current).getTime() || null;
+
+        const fieldDateValue = new Date(field.value).getTime() || null;
+
+        if (initialDateValue !== fieldDateValue) {
+            setRenderCount(r => r + 1);
+            parse
+                ? field.onChange(parse(field.value))
+                : field.onChange(field.value);
+            initialDefaultValueRef.current = field.value;
+        }
+    }, [setRenderCount, parse, field]);
+
+    const { onBlur: onBlurFromField } = field;
```

Comment:
Why have a different logic for `onBlur` and `onFocus`?

Reply:
react-hook-form only handle `onBlur` not `onFocus` and it will call any custom `onBlur` provided